File: python/mirage/core/redis/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.redis import RedisAccessor
from mirage.cache.index import IndexCacheStore, IndexEntry
from mirage.types import PathSpec
from mirage.utils.errors import enoent
from mirage.utils.key_prefix import mount_prefix_of
from mirage.utils.path import norm


async def readdir(
    accessor: RedisAccessor,
    path: PathSpec,
    index: IndexCacheStore,
) -> list[str]:
    virtual = path.virtual
    if isinstance(path, PathSpec):
        prefix = mount_prefix_of(path.virtual, path.resource_path)
        path = path.directory if path.pattern else path.virtual
    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    store = accessor.store
    # Canonical key: no trailing slash (except root), or the same dir
    # indexes under two keys and cache hits return doubled-slash entries.
    virtual_key = prefix + path if prefix else path
    virtual_key = virtual_key.rstrip("/") or "/"
    listing = await index.list_dir(virtual_key)
    if listing.entries is not None:
        return listing.entries
    p = norm(path)
    if not await store.has_dir(p):
        raise enoent(virtual)
    dir_prefix = p.rstrip("/") + "/"
    seen: set[str] = set()
    all_files = await store.list_files()
    all_dirs = await store.list_dirs()
    for key in all_files + list(all_dirs):
        if key == p:
            continue
        if key.startswith(dir_prefix):
            remainder = key[len(dir_prefix):]
            child = remainder.split("/")[0]
            if child:
                seen.add(dir_prefix + child)
    entries = sorted(seen)
    virtual_entries = sorted((prefix + e if prefix else e) for e in entries)
    index_entries = [(
        e.rsplit("/", 1)[-1],
        IndexEntry(
            id=e,
            name=e.rsplit("/", 1)[-1],
            resource_type="file",
        ),
    ) for e in entries]
    await index.set_dir(virtual_key, index_entries)
    return virtual_entries

File: typescript/packages/core/src/core/redis/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { readdirError } from '../../utils/errors.ts'
import { mountPrefixOf } from '../../utils/key_prefix.ts'
import { rstripSlash } from '../../utils/slash.ts'
import { compareCodePoints } from '../../utils/sort.ts'
import type { RedisAccessor } from '../../accessor/redis.ts'
import { RedisIndexEntry } from './entry.ts'
import { norm } from './utils.ts'

export async function readdir(
  accessor: RedisAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  // A pattern spec addresses the directory whose entries the glob filters,
  // and the rest of this function works in mount-relative space, so the
  // directory has to be read off `dir` rather than off the virtual
  // `directory` string (python strips the prefix by hand for the same
  // reason).
  const target = path.pattern !== null ? path.dir : path
  const virtual = target.mountPath
  const mountPrefix = mountPrefixOf(target.virtual, target.resourcePath)
  // Canonical key: no trailing slash (except root), or the same dir
  // indexes under two keys and cache hits return doubled-slash entries.
  const virtualKey = rstripSlash(mountPrefix + virtual) || '/'
  if (index !== undefined) {
    const cached = await index.listDir(virtualKey)
    if (cached.entries !== undefined && cached.entries !== null) {
      return cached.entries
    }
  }
  const store = accessor.store
  const p = norm(virtual)
  if (!(await store.hasDir(p))) {
    throw await readdirError(
      path,
      p,
      (k) => store.hasFile(k),
      (k) => store.hasDir(k),
    )
  }
  const dirPrefix = p === '/' ? '/' : `${p}/`
  const seen = new Set<string>()
  const files = await store.listFiles()
  const dirs = await store.listDirs()
  for (const key of [...files, ...dirs]) {
    if (key === p) continue
    if (!key.startsWith(dirPrefix)) continue
    const remainder = key.slice(dirPrefix.length)
    const first = remainder.split('/')[0]
    if (first !== undefined && first.length > 0) seen.add(dirPrefix + first)
  }
  const sorted = [...seen].sort(compareCodePoints)
  const virtualEntries = sorted.map((entry) => `${mountPrefix}${entry}`)
  if (index !== undefined) {
    const fileSet = new Set(files)
    const entries: [string, RedisIndexEntry][] = sorted.map((e) => {
      const name = e.slice(e.lastIndexOf('/') + 1)
      return [name, fileSet.has(e) ? RedisIndexEntry.file(e) : RedisIndexEntry.folder(e)]
    })
    await index.setDir(virtualKey, entries)
  }
  return virtualEntries
}
